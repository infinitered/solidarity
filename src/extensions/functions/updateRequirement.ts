module.exports = async (requirement, settings, context) => {

  const { head, tail, pipe, flatten, map } = require('ramda')
  const checkCLIForUpdates = require('./checkCLIForUpdates')
  const skipRule = require('./skipRule')
  
  const { print } = context
  const requirementName = head(requirement)
  const rules = pipe(tail, flatten)(requirement)

  let ruleString = ''
  const spinner = print.spin(`Updating ${requirementName}`)

  // check each rule for requirement
  const ruleChecks = await map(async (rule) => {
    // skip if we can't update
    if (skipRule(rule.platform) || !rule.semver) return []

    switch (rule.rule) {
      // Handle CLI rule update
      case 'cli':
        const updateResult = await checkCLIForUpdates(rule, context)
        ruleString = `Keep ${rule.binary} ${rule.semver}`
        if (updateResult) {
          spinner.succeed(updateResult)
          return updateResult
        } else {
          spinner.succeed(ruleString)
          return []
        }
      default:
        return []
    }
  }, rules)

  // Run all the rule checks for a requirement
  return Promise.all(ruleChecks)
    .then(results => {
      spinner.stop()
      return results
    })
    .catch(err => print.error(err))
}
