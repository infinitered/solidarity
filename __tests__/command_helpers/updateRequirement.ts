import updateRequirement from '../../src/extensions/functions/updateRequirement'

test('updateRequirement exists', () => expect(updateRequirement).toMatchSnapshot())
