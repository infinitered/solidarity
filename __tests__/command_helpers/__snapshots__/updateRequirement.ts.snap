// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateRequirement exists 1`] = `[Function]`;
